File: ios/Tests/PhpNitroProtocolTests/ScreenNavigationTests.swift
import XCTest
@testable import PhpNitroProtocol

final class ScreenNavigationTests: XCTestCase {
    func testNavigatePushesOntoTheStack() {
        let result = ScreenNavigation.reduce(action: "navigate:product?id=42", stack: ["home"])

        XCTAssertEqual(result, .fetch(stack: ["home", "product?id=42"], action: nil))
    }

    func testTabResetsTheWholeStack() {
        let result = ScreenNavigation.reduce(action: "tab:profile", stack: ["home", "product?id=42", "reviews"])

        XCTAssertEqual(result, .fetch(stack: ["profile"], action: nil))
    }

    func testBackPopsTheStackWhenMoreThanOneScreen() {
        let result = ScreenNavigation.reduce(action: "back", stack: ["home", "product?id=42"])

        XCTAssertEqual(result, .fetch(stack: ["home"], action: nil))
    }

    func testBackIsANoOpOnTheRootScreen() {
        let result = ScreenNavigation.reduce(action: "back", stack: ["home"])

        XCTAssertEqual(result, .fetch(stack: ["home"], action: nil))
    }

    func testClientTabIsFullyLocalWithNoFetch() {
        let result = ScreenNavigation.reduce(action: "clientTab:tabs1:2", stack: ["home"])

        XCTAssertEqual(result, .clientTabOnly(key: "tabs1", index: 2))
    }

    func testMalformedClientTabFallsBackToAPlainFetch() {
        let result = ScreenNavigation.reduce(action: "clientTab:tabs1", stack: ["home"])

        XCTAssertEqual(result, .fetch(stack: ["home"], action: nil))
    }

    func testAPlainActionRefetchesTheCurrentScreenWithIt() {
        let result = ScreenNavigation.reduce(action: "counter:increment", stack: ["home"])

        XCTAssertEqual(result, .fetch(stack: ["home"], action: "counter:increment"))
    }

    func testToggleWithMetaExtractsTheNextValueAsALocalFieldUpdate() {
        let result = ScreenNavigation.reduce(action: "toggle:agree", stack: ["home"], metaJson: #"{"next":"1"}"#)

        XCTAssertEqual(result, .fieldUpdate(key: "agree", value: "1"))
    }

    func testToggleWithAnEmptyNextIsStillAFieldUpdate() {
        // An unchecked Checkbox's own "next" is the empty string, not
        // absent — Checkbox.php's own docblock: `'next' => $checked ? ''
        // : '1'`.
        let result = ScreenNavigation.reduce(action: "toggle:agree", stack: ["home"], metaJson: #"{"next":""}"#)

        XCTAssertEqual(result, .fieldUpdate(key: "agree", value: ""))
    }

    func testToggleWithNoMetaFallsBackToAPlainFetch() {
        let result = ScreenNavigation.reduce(action: "toggle:agree", stack: ["home"])

        XCTAssertEqual(result, .fetch(stack: ["home"], action: "toggle:agree"))
    }

    func testToggleWithMalformedMetaFallsBackToAPlainFetch() {
        let result = ScreenNavigation.reduce(action: "toggle:agree", stack: ["home"], metaJson: "not json")

        XCTAssertEqual(result, .fetch(stack: ["home"], action: "toggle:agree"))
    }
}

File: windows/PhpNitroDesktop.Protocol.Tests/ScreenNavigationTests.cs
using PhpNitroDesktop.Protocol;
using Xunit;

namespace PhpNitroDesktop.Protocol.Tests;

// Ported case for case from ScreenNavigationTests.swift (iOS) /
// test_navigation.py (Linux).
public class ScreenNavigationTests
{
    [Fact]
    public void NavigatePushesOntoTheStack()
    {
        var result = ScreenNavigation.Reduce("navigate:product?id=42", new[] { "home" });

        var fetch = Assert.IsType<Fetch>(result);
        Assert.Equal(new[] { "home", "product?id=42" }, fetch.Stack);
        Assert.Null(fetch.Action);
    }

    [Fact]
    public void TabResetsTheWholeStack()
    {
        var result = ScreenNavigation.Reduce("tab:profile", new[] { "home", "product?id=42", "reviews" });

        var fetch = Assert.IsType<Fetch>(result);
        Assert.Equal(new[] { "profile" }, fetch.Stack);
    }

    [Fact]
    public void BackPopsTheStackWhenMoreThanOneScreen()
    {
        var result = ScreenNavigation.Reduce("back", new[] { "home", "product?id=42" });

        var fetch = Assert.IsType<Fetch>(result);
        Assert.Equal(new[] { "home" }, fetch.Stack);
    }

    [Fact]
    public void BackIsANoOpOnTheRootScreen()
    {
        var result = ScreenNavigation.Reduce("back", new[] { "home" });

        var fetch = Assert.IsType<Fetch>(result);
        Assert.Equal(new[] { "home" }, fetch.Stack);
    }

    [Fact]
    public void ClientTabIsFullyLocalWithNoFetch()
    {
        var result = ScreenNavigation.Reduce("clientTab:tabs1:2", new[] { "home" });

        var clientTabOnly = Assert.IsType<ClientTabOnly>(result);
        Assert.Equal("tabs1", clientTabOnly.Key);
        Assert.Equal(2, clientTabOnly.Index);
    }

    [Fact]
    public void MalformedClientTabFallsBackToAPlainFetch()
    {
        var result = ScreenNavigation.Reduce("clientTab:tabs1", new[] { "home" });

        var fetch = Assert.IsType<Fetch>(result);
        Assert.Equal(new[] { "home" }, fetch.Stack);
        Assert.Null(fetch.Action);
    }

    [Fact]
    public void APlainActionRefetchesTheCurrentScreenWithIt()
    {
        var result = ScreenNavigation.Reduce("counter:increment", new[] { "home" });

        var fetch = Assert.IsType<Fetch>(result);
        Assert.Equal(new[] { "home" }, fetch.Stack);
        Assert.Equal("counter:increment", fetch.Action);
    }

    [Fact]
    public void ToggleWithMetaExtractsTheNextValueAsALocalFieldUpdate()
    {
        var result = ScreenNavigation.Reduce("toggle:agree", new[] { "home" }, """{"next":"1"}""");

        var fieldUpdate = Assert.IsType<FieldUpdate>(result);
        Assert.Equal("agree", fieldUpdate.Key);
        Assert.Equal("1", fieldUpdate.Value);
    }

    [Fact]
    public void ToggleWithAnEmptyNextIsStillAFieldUpdate()
    {
        // An unchecked Checkbox's own "next" is the empty string, not
        // absent — Checkbox.php's own docblock: 'next' => $checked ? ''
        // : '1'.
        var result = ScreenNavigation.Reduce("toggle:agree", new[] { "home" }, """{"next":""}""");

        var fieldUpdate = Assert.IsType<FieldUpdate>(result);
        Assert.Equal("agree", fieldUpdate.Key);
        Assert.Equal("", fieldUpdate.Value);
    }

    [Fact]
    public void ToggleWithNoMetaFallsBackToAPlainFetch()
    {
        var result = ScreenNavigation.Reduce("toggle:agree", new[] { "home" });

        var fetch = Assert.IsType<Fetch>(result);
        Assert.Equal("toggle:agree", fetch.Action);
    }

    [Fact]
    public void ToggleWithMalformedMetaFallsBackToAPlainFetch()
    {
        var result = ScreenNavigation.Reduce("toggle:agree", new[] { "home" }, "not json");

        var fetch = Assert.IsType<Fetch>(result);
        Assert.Equal("toggle:agree", fetch.Action);
    }
}
